core/src/main/java/hudson/logging/LogRecorder.java

```diff
@@ -85,16 +87,60 @@
         return ts;
     }
 
+    @Restricted(NoExternalUse.class)
+    @VisibleForTesting
+    public static Set<String> getAutoCompletionCandidates(List<String> loggerNamesList) {
+        Set<String> loggerNames = new HashSet<>(loggerNamesList);
+
+        // now look for package prefixes that make sense to offer for autocompletion:
+        // Only prefixes that match multiple loggers will be shown.
+        // Example: 'org' will show 'org', because there's org.apache, org.jenkinsci, etc.
+        // 'io' might only show 'io.jenkins.plugins' rather than 'io' if all loggers starting with 'io' start with 'io.jenkins.plugins'.
+        HashMap<String, Integer> seenPrefixes = new HashMap<>();
+        SortedSet<String> relevantPrefixes = new TreeSet<>();
```

Comment:
Normally I would have suggested to use `new TreeSet<>(String.CASE_INSENSITIVE_ORDER)`, but the separation between the pure package name and the Logger name is somewhat nice. 👍 

I hope we will never have package with uppercased name or class with lowercased name.

Reply:
> I hope we will never have package with uppercased name or class with lowercased name. 

IIRC we have some such packages in the code, but I do not care much about such case. Performance is better with the standard ordering